## Loop Structure

When we write programs, we may encounter a scenario where we need to repeat a certain instruction or instructions. For example, we need to output "hello, world" on the screen every 1 second and continue to output it for one hour. The following code can complete this operation once. If we want to output it continuously for one hour, we need to write this code 3600 times. Are you willing to do this?

In [9]:
import time

print('hello, world')
time.sleep(1)

hello, world


> **Description**: The `sleep` function of the Python built-in `time` module can realize the sleep of the program. The parameter `1` represents the number of seconds of sleep. You can use `int` or `float` types, for example, `0.05` represents `50` milliseconds. We will explain the knowledge about functions and modules in subsequent courses.

To deal with the problems in the above scenario, we can use loop structures in Python programs. The so-called loop structure is a structure in the program that controls the repeated execution of one or more instructions. With such a structure, the code just now does not need to be written 3600 times, but written once and then put into the loop structure to repeat 3600 times. There are two ways to construct loop structures in Python language, one is the `for-in` loop, and the other is the `while` loop.

### for-in loop

If you know the number of times the loop is executed, we recommend using the `for-in` loop. For example, the scenario mentioned above where the loop is repeated 3,600 times can be implemented with the following code. Note that the code block controlled by the `for-in` loop is also constructed by indentation, which is the same as the method of constructing code blocks in the branch structure. The code block controlled by the `for-in` loop is called the loop body. Usually, the statements in the loop body will be executed repeatedly according to the loop settings.

In [13]:
import time

for i in range(3):
    print('hello, world')
    time.sleep(1)

hello, world
hello, world
hello, world
hello, world
hello, world
hello, world


KeyboardInterrupt: 

It should be noted that the `range(3600)` in the above code can construct a range from `0` to `3599`. When we put such a range into the `for-in` loop, we can use the previous loop variable `i` to sequentially take out integers from `0` to `3599`, which will cause the statements in the `for-in` code block to repeat 3600 times. Of course, the usage of `range` is very flexible. The following list gives an example of using the `range` function:

- `range(101)`: can be used to generate integers in the range of `0` to `100`. Note that `101` cannot be obtained.
- `range(1, 101)`: can be used to generate integers in the range of `1` to `100`, which is equivalent to the left-closed and right-open setting, that is, `[1, 101)`.
- `range(1, 101, 2)`: can be used to generate odd numbers from `1` to `100`, where `2` is the step size (span), that is, the value that increases each time, and `101` cannot be obtained.
- `range(100, 0, -2)`: can be used to generate even numbers from `100` to `1`, where `-2` is the step size (span), that is, the value that decreases each time, and `0` cannot be obtained.

You may have noticed that the output operation and the sleep operation in the above code do not use the loop variable `i`. For the `for-in` loop structure that does not require a loop variable, according to Python programming conventions, we usually name the loop variable `_`. The modified code is as follows. Although the result does not change much, writing it this way makes you look more professional and instantly more impressive.

In [ ]:
import time

for _ in range(3600):
    print('hello, world')
    time.sleep(1)

The above code will take one hour to execute. If you want to end the program early, you can click the Kernel and interrupt Kernel in Jupyter in top right menu, as shown in the figure below. If you run the code in the command prompt or terminal, you can use the key combination `ctrl+c` to terminate the program.

<img src="res/day06/terminate_program.png" style="zoom:40%;">

Next, we use the for-in loop to sum the integers from 1 to 100, that is, $\small{\sum_{n=1}^{100}{n}}$.

In [47]:
total = 0
for i in range(1, 101):
    total += i
print(total)

5050


In the above code, the variable `total` is used to store the accumulated result. During the loop, the value of the loop variable `i` will be taken from 1 to 100. For each value of the variable `i`, we executed `total += i`, which is equivalent to `total = total + i`. This statement implements the accumulation operation. Therefore, when the loop ends, we output the value of the variable `total`, and its value is 5050, the result of accumulating from 1 to 100. Note that there is no indentation before the statement `print(total)`. It is not controlled by the `for-in` loop and will not be executed repeatedly.

Let's write a code to sum the even numbers from 1 to 100, as shown below.

In [49]:
total = 0
for i in range(1, 101):
    if i % 2 == 0:
        total += i
print(total)

2550


> **Explanation**: In the above `for-in` loop, we use a branch structure to determine whether the loop variable `i` is an even number.

We can also modify the parameters of the `range` function, change the starting value and span to `2`, and use simpler code to achieve the sum of even numbers from 1 to 100.

In [ ]:
total = 0
for i in range(2, 101, 2):
    total += i
print(total)

Of course, a simpler way is to use Python's built-in `sum` function to sum, so we can save even the loop structure.

In [25]:
print(sum(range(2, 101, 2)))

2550


### while loop

If you want to construct a loop structure but you are not sure how many times the loop will repeat, we recommend using a while loop. The while loop controls the loop through a Boolean value or an expression that can produce a Boolean value. When the Boolean value or expression evaluates to True, the statements in the loop body (the code block with the same indentation under the while statement) are executed repeatedly. When the expression evaluates to False, the loop ends.

Next, we use the while loop to sum the integers from 1 to 100. The code is as follows.

In [27]:

"""
Sum of integers from 1 to 100

"""
total = 0
i = 1
while i <= 100:
    total += i
    i += 1
print(total)


5050


Compared to the `for-in` loop, in the above code, we add a variable `i` before the loop starts. We use this variable to control the loop, so the condition `i <= 100` is given after `while`. In the loop body of `while`, in addition to accumulation, we also need to increment the value of the variable `i`, so we add the statement `i += 1`, so that the value of `i` will be 1, 2, 3, ..., until 101. When `i` becomes 101, the condition of the `while` loop is no longer true, and the code will leave the `while` loop. At this time, we output the value of the variable `total`, which is the sum of 1 to 100, 5050.

If we want to sum even numbers from 1 to 100, we can modify the above code slightly.

In [31]:
total = 0
i = 2
while i <= 100:
    total += i
    i += 2
print(total)


2550


### break and continue

What happens if we set the condition of the while loop to True, that is, make the condition always true? Let's look at the following code, which still uses the while loop structure to calculate the sum of even numbers from 1 to 100.

In [ ]:
```python
"""
Sum of even numbers from 1 to 100

Version: 1.4
Author: Luo Hao
"""
total = 0
i = 2
while True:
    total += i
    i += 2
    if i > 100:
        break
print(total) 
```

In the above code, `while True` is used to construct a loop with a condition that is always true, which means that the loop will not end without special treatment. This is what we often call an "infinite loop". In order to stop the loop after the value of `i` exceeds 100, we use the `break` keyword, which terminates the execution of the loop structure. It should be noted that `break` can only terminate the loop it is in. This is something that needs to be paid attention to when using nested loop structures. We will talk about what a nested loop structure is later. In addition to `break`, there is another keyword `continue` that can be used in the loop structure. It can be used to abandon the subsequent code of this loop and directly let the loop enter the next round. The code is shown below.

In [33]:

"""
Sum of even numbers from 1 to 100

Version: 1.5
Author: Luo Hao
"""
total = 0
for i in range(1, 101):
    if i % 2 != 0:
        continue
    total += i
print(total)


2550


> **Note**: The code above uses the `continue` keyword to skip the case where `i` is an odd number. Only when `i` is an even number will it execute to `total += i`.

### Nested loop structure

Like branch structures, loop structures can also be nested, that is, loop structures can be constructed within loop structures. The following example demonstrates how to output a multiplication table (multiplication table) through nested loops.

In [37]:
"""
Print multiplication table

Version: 1.0
Author: Luo Hao
"""
for i in range(1, 10):
    for j in range(1, i + 1):
        print(f'{i}×{j}={i * j}', end='\t')
    print()

1×1=1	
2×1=2	2×2=4	
3×1=3	3×2=6	3×3=9	
4×1=4	4×2=8	4×3=12	4×4=16	
5×1=5	5×2=10	5×3=15	5×4=20	5×5=25	
6×1=6	6×2=12	6×3=18	6×4=24	6×5=30	6×6=36	
7×1=7	7×2=14	7×3=21	7×4=28	7×5=35	7×6=42	7×7=49	
8×1=8	8×2=16	8×3=24	8×4=32	8×5=40	8×6=48	8×7=56	8×8=64	
9×1=9	9×2=18	9×3=27	9×4=36	9×5=45	9×6=54	9×7=63	9×8=72	9×9=81	


In the above code, the for-in loop is used in the loop body of the for-in loop. The outer loop is used to control the output of the i-th row, while the inner loop is used to control the output of the j-th column in a row. Obviously, the output of the inner for-in loop is a whole row in the multiplication table. So when the inner loop is completed, we use a print() to achieve the effect of line break, so that the following output starts on a new line. The final output is as follows.

```
1×1=1	
2×1=2	2×2=4	
3×1=3	3×2=6	3×3=9	
4×1=4	4×2=8	4×3=12	4×4=16	
5×1=5	5×2=10	5×3=15	5×4=20	5×5=25	
6×1=6	6×2=12	6×3=18	6×4=24	6×5=30	6×6=36	
7×1=7	7×2=14	7×3=21	7×4=28	7×5=35	7×6=42	7×7=49	
8×1=8	8×2=16	8×3=24	8×4=32	8×5=40	8×6=48	8×7=56	8×8=64	
9×1=9	9×2=18	9×3=27	9×4=36	9×5=45	9×6=54	9×7=63	9×8=72	9×9=81
```

### Application of loop structure

#### Example 1: Determine prime numbers

Requirements: Enter a positive integer greater than 1 and determine whether it is a prime number.

> **Tip**: A prime number is an integer greater than 1 that can only be divided by 1 and itself. For example, for a positive integer $\small{n}$, we can determine whether it is a prime number by looking for factors of $\small{n}$ between 2 and $\small{n - 1}$. Of course, the loop does not have to start from 2 and end at $\small{n - 1}$, because for positive integers greater than 1, the factors should appear in pairs, so the loop can end when it reaches $\small{\sqrt{n}}$.

In [39]:
"""
Enter a positive integer greater than 1 to determine whether it is a prime number.

"""
num = int(input('Please enter a positive integer: '))
end = int(num ** 0.5)
is_prime = True
for i in range(2, end + 1):
    if num % i == 0:
        is_prime = False
        break
if is_prime:
    print(f'{num} is a prime number')
else:
    print(f'{num} is not a prime number')

Please enter a positive integer:  9899


9899 is not a prime number


> **Explanation**: In the above code, we use the Boolean variable `is_prime`. We first assign it to `True`, assuming that `num` is a prime number; next, we look for factors of `num` in the range of 2 to `num ** 0.5`. If factors of `num` are found, then it must not be a prime number. At this time, we assign `is_prime` to `False` and use the `break` keyword to terminate the loop structure; finally, we give different outputs depending on whether the value of `is_prime` is `True` or `False`.

#### Example 2: Greatest common divisor

Requirements: Enter two positive integers greater than 0 and find the greatest common divisor of the two numbers.

> **Hint**: The greatest common divisor of two numbers is the largest number among the common factors of the two numbers.

In [41]:
"""
Enter two positive integers to find their greatest common divisor

"""
x = int(input('x = '))
y = int(input('y = '))
for i in range(x, 0, -1):
    if x % i == 0 and y % i == 0:
        print(f'Greatest common divisor: {i}')
        break

x =  98
y =  98


Greatest common divisor: 98


> **Explanation**: In the above code, the loop variable values ​​of the `for-in` loop are from large to small, so the factor `i` that we find can divide both `x` and `y` is the greatest common divisor of `x` and `y`, and we use `break` to terminate the loop. If `x` and `y` are coprime, the loop will execute until `i` becomes 1, because 1 is a factor of all positive integers, and the greatest common divisor of `x` and `y` is 1.

The above code is not efficient in finding the greatest common divisor. If the value of `x` is `999999999998` and the value of `y` is `9999999999999`, it is obvious that the two numbers are coprime and the greatest common divisor is 1. However, if we use the above code, the loop will repeat `9999999999998` times, which is usually unacceptable. We can use the Euclidean algorithm to find the greatest common divisor, which can help us get the desired result faster. The code is shown below.

In [45]:

"""
Enter two positive integers to find their greatest common divisor

Version: 1.1
Author: Luo Hao
"""
x = int(input('x = '))
y = int(input('y = '))
while y % x != 0:
    x, y = y % x, x
print(f'Greatest common divisor: {x}')


x =  16
y =  4


Greatest common divisor: 4


> **Explanation**: The methods and steps to solve a problem can be called an algorithm. For the same problem, we can design different algorithms. Different algorithms will have differences in storage space usage and execution efficiency, and these differences represent the advantages and disadvantages of the algorithms. You can compare the above two paragraphs to understand why we say that the Euclidean algorithm is a better choice. In the above code, the statement `x, y = y % x, x` means assigning the value of `y % x` to `x`, and assigning the original value of `x` to `y`.

#### Example 3: Guess the number game

Requirements: The computer generates a random number between 1 and 100, and the player enters the number he or she guesses. The computer gives the corresponding prompt information "bigger", "smaller" or "correct guess". If the player guesses the number correctly, the computer prompts the user how many times he or she has guessed, and the game ends. Otherwise, the game continues.

```python
"""
Guess the number game

Version: 1.0
Author: Luo Hao
"""
import random

answer = random.randrange(1, 101)
counter = 0
while True:
    counter += 1
    num = int(input('Please enter: '))
    if num < answer:
        print('Bigger.')
    elif num > answer:
        print('Smaller.')
    else:
        print('Guessed it right.')
        break
print(f'You guessed {counter} times in total.')
```

> **Explanation**: The above code uses `import random` to import the `random` module of the Python standard library. The `randrange` function of this module helps us generate random numbers in the range of 1 to 100 (excluding 100). The variable `counter` is used to record the number of times the loop is executed, that is, how many times the user guessed in total. The value of `counter` will increase by 1 every time the loop is executed.

### Summarize

After learning the branching and looping structures in Python, we can solve many practical problems. Through this lesson, everyone should have learned that the `for` and `while` keywords can be used to construct loop structures. **If we know the number of times the loop structure is repeated in advance, we usually use **`for`** loop**; **If the number of times the loop structure is repeated is uncertain, we can use **`while`** loop**. In addition, we can **use **`break`** to terminate the loop** in the loop structure, **and we can also use **`continue`** keywords in the loop structure to let the loop structure directly enter the next round**.